kong/clustering/version_negotiation/init.lua

```diff
@@ -0,0 +1,340 @@
+
+local cjson = require "cjson.safe"
+local pl_file = require "pl.file"
+local http = require "resty.http"
+
+local constants = require "kong.constants"
+local clustering_utils = require "kong.clustering.utils"
+
+local ngx = ngx
+local ngx_log = ngx.log
+local ngx_ERR = ngx.ERR
+local ngx_DEBUG = ngx.DEBUG
+local _log_prefix = "[version-negotiation] "
+
+local KONG_VERSION
+local CLUSTERING_SYNC_STATUS = constants.CLUSTERING_SYNC_STATUS
+
+
+local _M = {}
+
+
+local function get_body()
+  ngx.req.read_body()
+  local body = ngx.req.get_body_data()
+  if body then
+    return body
+  end
+
+  local fname = ngx.req.get_body_file()
+  if fname then
+    return pl_file.read(fname)
+  end
+
+  return ""
+end
+
+local function response(status, body)
+  ngx.status = status
+
+  if type(body) == "table" then
+    ngx.header["Content-Type"] = "application/json"
+    body = cjson.encode(body)
+  end
+
+  ngx.say(body)
+  return ngx.exit(status)
+end
+
+local function response_err(msg)
+  return response(400, { message = msg })
+end
+
+local function verify_request(body)
+  if type(body.node) ~= "table" then
+    return false, "field \"node\" must be an object."
+  end
+
+  if type(body.node.id) ~= "string" then
+    return false, "field \"node.id\" must be a string."
+  end
+
+  if type(body.node.type) ~= "string" then
+    return false, "field \"node.type\" must be a string."
+  end
+
+  if type(body.node.version) ~= "string" then
+    return false, "field \"node.version\" must be a string."
+  end
+
+  if type(body.services_requested) ~= "table" then
+    return false, "field \"services_requested\" must be an array."
+  end
+
+  return true
+end
+
+
+local function node_info()
+  return {
+    id = kong.node.get_id()
+  }
+end
+
+local function cp_priority(name, req_versions, known_versions)
```

Comment:
There's a lot of business logic encoded in this module, and from the perspective of somebody who is not already well-versed in the version negotiation flow, I'm finding myself bouncing around the file a lot re-reading code to fully grok things. The code isn't particularly opaque or anything; it's just that I'm finding myself wondering:

> is the behavior in function 'foo' part of some specification/API/contract, or is it just an implementation detail or incidental thing that I'm not meant to care about?

Let's take this `cp_priority` function, for instance. Based on following the behavior of the function line-by-line I _think_ I am meant to understand that the ordering of the table in `kong.clustering.version_negotiation.services_known` is a significant piece of the business logic and determines which version is ultimately utilized, since we return on the first found item in the table. However, because there's not really an explanation or test case for this behavior, I'm not 100% confident in that I've made the right conclusion about this. All that's to say that I would be pretty nervous about touching this code if I needed to fix a bug or add a feature.

I think some comments and doc strings that help to outline/explain the process would be helpful (even if it's just to link out to an ADR or something). Also, while I _think_ the integration tests exercise most--if not all--of this code, the existence of some unit cases that specifically target the negotiation business logic (however simple) would be very helpful to future maintainers.

Reply:
yes, the version priority is determined by the CP only; and currently the way to set it is by the ordering of version records in the `services_know.lua` file.

Reply:
as for unit tests... there were some!   i was surprised not to find them anymore; I removed them when they became obsolete on a rereading of the spec :grin: and apparently forgot to rewrite them.  will add some